Sources/UIKit/PortalCollectionView.swift

```diff
@@ -0,0 +1,162 @@
+    fileprivate var selected: Int = 0
+    fileprivate var lastOffset: CGFloat = 0
+    
+    public init(items: [CollectionItemProperties<MessageType>], layoutEngine: LayoutEngine, layoutValues: CollectionViewLayoutValues) {
+        self.items = items
+        self.layoutEngine = layoutEngine
+        let layout = createFlowLayout(layoutValues: layoutValues)
+        super.init(frame: .zero, collectionViewLayout: layout)
+   
+        self.dataSource = self
+        self.delegate = self
+        
+        let identifiers = Array(Set(items.map { $0.identifier }))
+        identifiers.forEach { register(PortalCollectionViewCell<MessageType>.self, forCellWithReuseIdentifier: $0) }
+    }
+    
+    required public init?(coder aDecoder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+    
+    public func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
+        return items.count
+    }
+    
+    public func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
+        let item = items[indexPath.row]
+        let cell = dequeueReusableCell(with: item.identifier, for: indexPath)
+        cell.component = itemRender(at: indexPath)
+        cell.isDebugModeEnabled = isDebugModeEnabled
+        cell.render()
+        
+        return cell
+    }
+    
+    public func numberOfSections(in collectionView: UICollectionView) -> Int {
+        return 1
+    }
+    
+    public func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
+        let item = items[indexPath.row]
+        item.onTap |> { mailbox.dispatch(message: $0) }
+    }
+    
+    public func scrollViewWillBeginDragging(_ scrollView: UIScrollView) {
+        lastOffset = scrollView.contentOffset.x
+    }
+    
+    public func scrollViewWillEndDragging(_ scrollView: UIScrollView, withVelocity velocity: CGPoint,
+                                          targetContentOffset: UnsafeMutablePointer<CGPoint>) {
+        guard isSnapToCellEnabled else { return }
+        
+        let currentOffset = CGFloat(scrollView.contentOffset.x)
+        
+        if currentOffset == lastOffset {
+            return
+        }
+        
+        let lastPosition = selected
+        if currentOffset > lastOffset {
+            if lastPosition < items.count - 1 {
+                selected = lastPosition + 1 // Move to the right
+                scrollToItem(selected, animated: true)
+            }
+        } else if currentOffset < lastOffset {
+            if lastPosition >= 1 {
+                selected = lastPosition - 1
+                scrollToItem(selected, animated: true) // Move to the left
+            }
+        }
+    }
+    
+}
+
+fileprivate extension PortalCollectionView {
+    
+    fileprivate func scrollToItem(_ position: Int, animated: Bool) {
+        DispatchQueue.main.async(execute: {
```

Comment:
Could you add a comment explaining why do you need to dispatch this block of code